Lockfiles without version ranges

mattfarina opened the pull request:
There were two situations where a lockfile could end up with a version range as a version instead of a pinned version.

1. When `file://` was used. This would point to the file location on the filesystem but, instead of using the version in that chart at that file location the version range would be used.
2. If a repository was used that was not added with `helm repo add`. Instead of causing an error that the repository was not available the version range was used in the lockfile. This is not locking.

This change looks to correct these problems. The `file://` handling fix was fairly straight forward. In that case update gets the current version of the dependency on disk and locks to that.

Handling unknown repos went through 4 different major versions. This 4th one is the one that settled on because all the changes are in the download manager which handles network issues. Other cases had added network operations to the resolver which is meant to iterate over existing versions. I chose this as the network and downloading stay in the same struct and it's methods are largely reusable within the new logic.

Closes #8449 

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

mattfarina committed:
Locking file URIs to a version in lockfile

Previously, if a range was specified for a file:// url as a
dependency the range would be put in the lockfile. Lockfiles are
designed to pin to a specific version and not support ranges. This
is for reproducibility.

The change here pins to a the specific version of the chart
specified using the file:// when update is run.

Signed-off-by: Matt Farina <matt@mattfarina.com>

mattfarina committed:
Make unmanaged repositories version resolvable

If a repository was not know to helm (e.g. added using helm repo add)
then Helm would use the range set in the depenencies as the version
in the lock file. Lock files should not have ranges since they are
locked to versions.

Helm did this because the version information for repositories was
not know to Helm. This change fixes that by making the repository
and chart information known to Helm so it can resolve the versions.

Closes #8449

Signed-off-by: Matt Farina <matt@mattfarina.com>

mattfarina commented:
While this is technically a bug fix it does change some behavior when `helm dep build` is used. So, I'm putting it into v3.4.0.

mattfarina commented:
Note, this doesn't change the download mechanism so it will be compatible with lockfiles that had a version range slip in.

technosophos commented:
Say a chart dependency references repo `https://foo.example.com/stable`. My understanding is that in the current code, it will add a repository named something like `helm-manager-a576fe4...`. So if a user does a `helm repo list`, they'll see a really weirdly named repo.

Could we go for a more clear name, like naming it `foo.example.com/stable` or `foo-example-com-stable`? That might make it a little easier on the user.

mattfarina commented:
@technosophos the repo being added is just to the cache and the internals during update/build. The repository configuration isn't updated so `helm repo list` won't have the unmanaged one. The reason for the hash as a key name is so that the cache is highly unlikely to have collisions.

mattfarina committed:
Adding helm v4 todo

Signed-off-by: Matt Farina <matt@mattfarina.com>

technosophos reviewed (approved):
LGTM. Also paired with @mattfarina to ask questions, and I believe this is the best path forward for solving several related problems without breaking compatibility.

jdolitsky commented on pkg/downloader/manager.go:
```diff
@@ -147,14 +150,27 @@ func (m *Manager) Update() error {
+	// has some information about them and, when possible, the index files
+	// locally.
+	// TODO(mattfarina): Repositories should be explicitly added by end users
+	// rather than automattic. In Helm v4 require users to add repositories. They
```
typos in comment (automattic), "For security" vs. ", for securirty"

mattfarina replied:
Fixed... did you notice you misspelled "securirty" ;-). I find it amusing that a typo fix includes a typo.

jdolitsky replied:
hahaha

jdolitsky commented on pkg/downloader/manager.go:
```diff
@@ -525,15 +565,27 @@ func (m *Manager) parallelRepoUpdate(repos []*repo.Entry) error {
-				fmt.Fprintf(m.Out, "...Unable to get an update from the %q chart repository (%s):\n\t%s\n", r.Config.Name, r.Config.URL, err)
+				// For those dependencies that are not known to helm and using a
+				// generated key name we display the repo url.
+				if strings.HasPrefix(r.Config.Name, "helm-manager-") {
```
Should we make "helm-manager-" a constant, as its used here 3+ times?

mattfarina replied:
done

mattfarina committed:
Restoring Build behavior

Two things changed in this commit...

1. The Build behavior was restored and the change only impacts
   Update. This is a more minimal functionality change thats
   a more secure behavior
2. Cleanup from Josh's feedback on the PR to create a const
   and comment changes

Signed-off-by: Matt Farina <matt@mattfarina.com>

mattfarina commented:
@technosophos @jdolitsky I updated the PR per Josh's feedback and to undo a change I made to Build. Build keeps its behavior.

mattfarina commented:
Also, rebased off master due to a change landing there that impacted the same code this does.
